File: typescript/src/workshop/m4/s11_encapsulatecollection/start/Booking.ts
import { Money } from '../../../shared/Money.js';
import type { Seat } from '../Seat.js';

const PRICE_2D = Money.of('25.00');
const VIP_SURCHARGE = Money.of('10.00');

/**
 * Start: publiczna, mutowalna lista miejsc. `readonly` blokuje tylko przypisanie -
 * każdy może dodać, usunąć albo wyczyścić miejsca z pominięciem właściciela.
 */
export class Booking {
  readonly seats: Seat[] = [];

  total(): Money {
    let total = Money.ZERO;
    for (const seat of this.seats) {
      total = total.plus(seat.row >= 10 ? PRICE_2D.plus(VIP_SURCHARGE) : PRICE_2D);
    }
    return total;
  }
}

File: csharp/src/Training.Workshop/M4/S11EncapsulateCollection/Start/Booking.cs
namespace Training.Workshop.M4.S11EncapsulateCollection.Start;

/// <summary>
/// Start: publiczna, mutowalna lista miejsc. <c>readonly</c> blokuje tylko przypisanie -
/// każdy może dodać, usunąć albo wyczyścić miejsca z pominięciem właściciela.
/// </summary>
public sealed class Booking
{
    private const decimal Price2D = 25.00m;
    private const decimal VipSurcharge = 10.00m;

    public readonly List<Seat> Seats = [];

    public decimal Total()
    {
        decimal total = 0.00m;
        foreach (var seat in Seats)
        {
            total = total + (seat.Row >= 10 ? Price2D + VipSurcharge : Price2D);
        }
        return total;
    }
}
